### DISCUSSION TITLE
Should we crowdfund the developer so they can buy a newer iphone to make the library support iOS better?

### DISCUSSION THREAD

**Discussion by mrhempman69** [pschatzmann/ESP32-A2DP#495] [Q&A] [answered]
I think all 4 of the issues in the repository could be solved if the maintainer had a more modern iphone. 

I would donate $100 for the developer to buy a used iPhone (like SE 2020 or iphone 6s) so that they could fix those issues.

But I'm not sure that it would cover the time necessary for them to do it. For example, I could fix it myself too, but it would take me a lot longer, and I got other things to do.

Does anyone else want to chip in? Or would the maintainer do it for $100? I will contribute through their PayPal if needbe

Tagging these issues:
https://github.com/pschatzmann/ESP32-A2DP/issues/493
https://github.com/pschatzmann/ESP32-A2DP/issues/485
https://github.com/pschatzmann/ESP32-A2DP/issues/388
https://github.com/pschatzmann/ESP32-A2DP/issues/163

**Comment by pschatzmann** [comment]
Are you sure that the issue is in this project and not the basic ESP32 API implementation or in the IPhone basic API or IPhone Applications?
So the first thing that would need to be done is to check with the original [ESP32 demo application](https://github.com/espressif/esp-idf/tree/master/examples/bluetooth/bluedroid/classic_bt/a2dp_sink) and if it can be reproduced to file an issue with Espressif

This is an open source project, so I will be more then happy to accept any pull requests for this issue...

**Reply by mrhempman69** [reply]
Well I will try the ESP32 demo application, but remember
1) It works on Android and not iOS (unlikely to be ESP32 API issue but still possible)
2&3) Regular bluetooth headphones from e.g. chinese manufacturers work fine with these apps

**Comment by pschatzmann** [comment]
It could well be that there are some unhandled events (only used by apple), but you would see that as warnings in the log.

**Comment by mrhempman69** [comment]
Ok first pass, the ESP-IDF is broken for both android and iOS. Many ringbuffer overflows/underflows. Music is choppy. Volume changes are delayed, or can work for a bit (Android), but then become delayed as more streams go on...

I did a second pass using the menuconfig changes from the wiki, namely removing BLE... Did not help....

I'll have to look into this more tomorrow

**Comment by pschatzmann** [comment]
Not sure if this is matching your isssue: https://github.com/espressif/esp-idf/issues/6712
You can change the queue size by calling set_event_queue_size(size), thought I have increased the default limit to 20 a long time ago 
Try to test with disabled logs!
It might be also worth to check if increasing the i2s buffer size/count helps...

Please also notice that the example seems to contain a task for volume_change_simulation which need to be disabled...

**Comment by pschatzmann** [comment]
I just discovered that it might make sense to include the latest changes to the i2s output task which might help in this case. So the BluetoothA2DPSinkQueued would need to be extended a bit to include the logic related to ringbuffer_mode

**Comment by pschatzmann** [comment, marked as answer]
I committed a new version of BluetoothA2DPSinkQueued which might resolve your issues...
You will need to run your sketch with this class instead of the simple BluetoothA2DPSink!
